src/common/heartbeat2/src/server.cpp

```diff
@@ -0,0 +1,94 @@
+#include <chrono>
+#include <memory>
+#include <rclcpp/rclcpp.hpp>
+
+#include "csm_common_interfaces/msg/e_stop.hpp"
+#include "csm_common_interfaces/srv/heartbeat.hpp"
+
+class Server : public rclcpp::Node {
+public:
+    Server() : Node("heartbeat_producer") {
+        // service to receive pings
+        heartbeatSrv = this->create_service<csm_common_interfaces::srv::Heartbeat>("heartbeat_srv", std::bind(&Server::heartbeat_callback, this, std::placeholders::_1, std::placeholders::_2));
+
+        // Setting up handshake
+        RCLCPP_INFO(this->get_logger(), "Waiting for ROV");
+        handshakeSrv = this->create_service<csm_common_interfaces::srv::Heartbeat>("handshake", std::bind(&Server::handshake, this, std::placeholders::_1, std::placeholders::_2));
+
+        // create timers
+        watchdog = this->create_wall_timer(std::chrono::milliseconds(300), std::bind(&Server::watchdog_callback, this));
+        watchdog->cancel();
+
+        // create estop publisher
+        estopPub = this->create_publisher<csm_common_interfaces::msg::EStop>("estop", 10);
+    }
+
+private:
+    void handshake(csm_common_interfaces::srv::Heartbeat::Request::SharedPtr request, csm_common_interfaces::srv::Heartbeat::Response::SharedPtr response) {
+        // send response to handshake request to start heartbeat
+        unsigned int millisec = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+        response->milliseconds = request->milliseconds;
+        _lastPingTime = millisec;
```

Comment:
check that last ping time is not greater than millisec in case this handshake arrives late